### PR TITLE
Login, replist, and some feed functionality

### PR DESCRIPTION
**Pull request by jindaln**
Made a login page in index.html, created repList.html, added servlets repListServlet, repAnswerServlet, newPostServlet, and repInDatastoreServlet. Abstracted out helper functions into java classes with static methods - Parse.java, QueryDatastore.java, and InsertandUpdate.java. Some javascript functionality for fetching repListServlet and displaying the list of reps and only linking rep's name to associated feed if rep has made account and is in datastore

### CODE REVIEW

File: src/main/java/com/google/sps/data/InsertAndUpdate.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    private static final String REP_POSTS = "Posts";
+
+    private static final String POST_ENTITY_TYPE = "Post";
+    private static final String POST_QUESTION = "Question";
+    private static final String POST_ANSWER = "Answer";
+    private static final String POST_REPLIES = "Replies";
+
+    private static final String COMMENT_ENTITY_TYPE = "Comment";
+    private static final String COMMENT_NAME = "Nick Name";
+    private static final String COMMENT_MSG = "Message";
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/data/InsertAndUpdate.java:30]: I moved all the constants to a separate class in my PR. How should we avoid the tons and tons of merge conflicts we are bound to have?  I also had to change some of the functions when testing and change variables to camel-case and stuff. I can send you the file or you can remove from PR and it'll be fine when we merge

File: src/main/java/com/google/sps/data/tests/testQueryForRepresentatives.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
+// import static com.google.appengine.api.datastore.FetchOptions.Builder.withLimit;
+// import static org.junit.Assert.assertEquals;
+
+// import com.google.appengine.api.datastore.DatastoreService;
+// import com.google.appengine.api.datastore.DatastoreServiceFactory;
+// import com.google.appengine.api.datastore.Entity;
+// import com.google.appengine.api.datastore.EntityNotFoundException;
+// import com.google.appengine.api.datastore.PreparedQuery;
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/data/tests/testQueryForRepresentatives.java:8]: just remove from PR cuz I have it in mine

File: src/main/java/com/google/sps/servlets/NewPostServlet.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            InsertAndUpdate.updateRepresentativePostList(postId, repId);
+        } 
+        catch(EntityNotFoundException e) {
+            System.out.println("Unable to update representative post list"); 
+            System.exit(0);
+        }
+
+        response.sendRedirect("feed.html");
+    }
+}
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/servlets/NewPostServlet.java:61]: trailing new line

File: src/main/java/com/google/sps/servlets/RepAnswerServlet.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import com.google.appengine.api.datastore.Key; 
+import com.google.appengine.api.datastore.KeyFactory;
+
+@WebServlet ("/rep_answer")
+public class RepAnswerServlet extends HttpServlet{
+    private static final String POST_ENTITY_TYPE = "Post";
+    private static final String POST_QUESTION = "Question";
+    private static final String POST_ANSWER = "Answer";
+    private static final String POST_REPLIES = "Replies";
+
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/servlets/RepAnswerServlet.java:25]: I made a constants class so just import that. (it's called Constants.java so all of these become Constants.POST_ANSWER, etc. )

File: src/main/java/com/google/sps/servlets/RepAnswerServlet.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        try {
+            post = QueryDatastore.queryForPost(postId);
+        } 
+        catch(EntityNotFoundException e) {
+            System.out.println("Unable to query for post"); 
+            System.exit(0);
+        }
+        post.setProperty(POST_ANSWER, commentId);
+    }
+}
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/servlets/RepAnswerServlet.java:43]: trailing new line

File: src/main/java/com/google/sps/servlets/RepInDatastoreServlet.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        String repName = request.getParameter("repName");
+        Entity representative = QueryDatastore.queryForRepresentative(repName);
+        boolean output = false;
+        if (representative != null){
+            output = true;
+        }
+        response.setContentType("text/html");
+        response.getWriter().println(output);
+    }
+}
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/servlets/RepInDatastoreServlet.java:25]: trailing new line

File: src/main/java/com/google/sps/servlets/RepInDatastoreServlet.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException{
+        String repName = request.getParameter("repName");
+        Entity representative = QueryDatastore.queryForRepresentative(repName);
+        boolean output = false;
+        if (representative != null){
+            output = true;
+        }
+        response.setContentType("text/html");
+        response.getWriter().println(output);
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/servlets/RepInDatastoreServlet.java:23]: do response.getWriter().println(representative != null);

File: src/main/java/com/google/sps/servlets/RepListServlet.java [hunk 8]
```diff
@@ -5,11 +5,66 @@
...
+
+        URIBuilder builder = new URIBuilder();
+        builder.setScheme("https").setHost("www.googleapis.com/civicinfo/v2/representatives")
+        .setParameter("key", API_KEY)
+        .setParameter("address", zipcode);
+        URI uri = null;
+        try{
+            uri = builder.build();
+        } catch(URISyntaxException e){
+            logger.error(e);
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/servlets/RepListServlet.java:42]: exit the program after logging the error

File: src/main/java/com/google/sps/servlets/RepListServlet.java [hunk 9]
```diff
@@ -5,11 +5,66 @@
...
+        HttpResponse httpresponse = null;
+        String responseString = null;
+
+        try {
+            httpresponse = httpclient.execute(httpget);
+        } catch (IOException e) {
+            logger.error(e);
+        } catch (Exception e) {
+            logger.error(e);
+        }
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/servlets/RepListServlet.java:55]: exit program after logging errors

File: src/main/webapp/script.js [hunk 10]
```diff
@@ -12,17 +12,60 @@
...
+}
+
+function storeZipCodeAndNickname(){
+    event.preventDefault();
+    var nickname = document.getElementById("nickname").value;
+    var zipcode = document.getElementById("zipcode").value;
+    localStorage.setItem("nickname", nickname);
+    localStorage.setItem("zipcode", zipcode);
+    window.location.href = "/repList.html";
+
```
Comment (swetha-gangu) [inline src/main/webapp/script.js:35]: remove empty space

File: src/main/java/com/google/sps/data/InsertAndUpdate.java (hunk 1 above)
Comment (jindaln) [reply src/main/java/com/google/sps/data/InsertAndUpdate.java:30]: I honestly don't think the merge will be too bad - we'll just keep your version but either way, you can send file and I'll update if we have time before we integrate

File: src/main/java/com/google/sps/data/tests/testQueryForRepresentatives.java (hunk 2 above)
Comment (jindaln) [reply src/main/java/com/google/sps/data/tests/testQueryForRepresentatives.java:8]: yep yep

File: src/main/java/com/google/sps/servlets/NewPostServlet.java (hunk 3 above)
Comment (jindaln) [reply src/main/java/com/google/sps/servlets/NewPostServlet.java:61]: this is weird - it shows I have a trailing line in my cloud shell?

File: src/main/java/com/google/sps/servlets/RepAnswerServlet.java (hunk 4 above)
Comment (jindaln) [reply src/main/java/com/google/sps/servlets/RepAnswerServlet.java:25]: just added Constants and adjusted everything - hopefully should minimize conflicts

File: src/main/java/com/google/sps/servlets/RepInDatastoreServlet.java (hunk 6 above)
Comment (jindaln) [reply src/main/java/com/google/sps/servlets/RepInDatastoreServlet.java:25]: not sure why its not showing up :(

File: src/main/java/com/google/sps/data/Constants.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+    public static final String POST_ENTITY_TYPE = "Post";
+    public static final String POST_QUESTION = "Question";
+    public static final String POST_ANSWER = "Answer";
+    public static final String POST_REPLIES = "Replies";
+
+    public static final String COMMENT_ENTITY_TYPE = "Comment";
+    public static final String COMMENT_NAME = "Nick Name";
+    public static final String COMMENT_MSG = "Message";
+}
```
Comment (jynCoder) [inline src/main/java/com/google/sps/data/Constants.java:18]: nit: Don't forget trailing newlines at the end of files

File: src/main/java/com/google/sps/servlets/RepInDatastoreServlet.java [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+@WebServlet ("/rep_in_datastore")
+public class RepInDatastoreServlet extends HttpServlet{
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException{
+        String repName = request.getParameter("repName");
+        Entity representative = QueryDatastore.queryForRepresentative(repName);
+        response.setContentType("text/html");
+        response.getWriter().println(representative != null);
+
```
Comment (swetha-gangu) [inline src/main/java/com/google/sps/servlets/RepInDatastoreServlet.java:20]: remove empty line

File: src/main/java/com/google/sps/servlets/NewPostServlet.java [hunk 13]
```diff
@@ -18,12 +18,19 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import javax.servlet.ServletException;
+
+/*
```
Comment (epoms) [inline src/main/java/com/google/sps/servlets/NewPostServlet.java:23]: Make it a doc comment. javadocs are a specific type of comment but the distinction is important, automated systems can read and expose them differently. (literally just add another * to the opening)

https://www.tutorialspoint.com/java/java_documentation.htm

File: src/main/java/com/google/sps/data/Comment.java [hunk 14]
```diff
@@ -3,9 +3,39 @@
...
+        return comment; 
+    }
+
+    public long getID(){
+        return id; 
+    }
+
+    @Override
+    public boolean equals(Object o) {   
+        if (o == this) {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:28]: I don't know if I made the comment about equals on this PR or another but:

o==this block is not necessary, please remove it.

typically we do 

Comment that = (Comment) o;
return this.getComment().equals(that.getComment())
           && this.getDisplayName().equals(that.getDisplayName());

(use the this/that pattern).

You'll notice that you have to use the equals methods to test equality for Strings.

Also why does this not test the ID?

File: src/main/java/com/google/sps/data/Comment.java [hunk 15]
```diff
@@ -3,9 +3,39 @@
...
+
+    public long getID(){
+        return id; 
+    }
+
+    @Override
+    public boolean equals(Object o) {   
+        if (o == this) { 
+            return true; 
+        }
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:30]: you must implement hashCode() if you implement equals.

File: src/main/java/com/google/sps/data/Comment.java [hunk 16]
```diff
@@ -3,9 +3,39 @@
 public class Comment{
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:3]: make the class final?

File: src/main/java/com/google/sps/data/Constants.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
+package com.google.sps.data;
+
+public final class Constants {
+    private Constants() {}
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Constants.java:4]: move the constructor to the bottom of the class since it doesn't contain real logic.

File: src/main/java/com/google/sps/data/InsertAndUpdate.java [hunk 18]
```diff
@@ -0,0 +1,53 @@
+package com.google.sps.data;
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/InsertAndUpdate.java:1]: I had a bunch of comments about these classes on the other PR. please address them.

File: src/main/java/com/google/sps/data/Post.java [hunk 19]
```diff
@@ -1,15 +1,50 @@
 package com.google.sps.data;
 
+import com.google.sps.data.Comment;
 import java.util.List;
 
 public class Post{
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Post.java:6]: mark final

File: src/main/java/com/google/sps/data/Post.java [hunk 20]
```diff
@@ -1,15 +1,50 @@
...
+    public Comment getQuestion(){
+        return question; 
+    }
+
+    public Comment getAnswer(){
+        return answer; 
+    }
+
+    public List<Comment> getReplies(){
+        return replies;
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Post.java:28]: you want to copy this list so no one else changes this out from under you.

File: src/main/java/com/google/sps/data/Post.java [hunk 21]
```diff
@@ -1,15 +1,50 @@
...
+        return replies; 
+    }
+
+    public long getID() {
+        return id; 
+    }
+
+    @Override
+    public boolean equals(Object o) {   
+        if (o == this) {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Post.java:37]: all the comments about equals above also apply here.

Comment (v1shanker) [reply src/main/java/com/google/sps/data/Post.java:37]: including hashCode

File: src/main/java/com/google/sps/data/Representative.java [hunk 22]
```diff
@@ -6,11 +6,45 @@
 public class Representative{
     private final String name;
     private final String title;
-    private final List<Post> replies;
+    private final List<Post> posts;
+    private final long id; 
 
-    public Representative(String name, String title, List<Post> replies){
+    public Representative(String name, String title, List<Post> posts, long id){
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Representative.java:12]: see all the comments on the other POJO classes.

File: src/main/java/com/google/sps/servlets/NewPostServlet.java [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+        String comment = request.getParameter("comment");
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        Entity repEntity = QueryDatastore.queryForRepresentative(repName); 
+        Representative rep = null; 
+        try {
+            rep = Parse.parseRepresentative(repEntity); 
+        } 
+        catch(EntityNotFoundException e) {
+            System.out.println("Unable to parse representative from datastore");
```
Comment (v1shanker) [inline src/main/java/com/google/sps/servlets/NewPostServlet.java:45]: use the logger and get rid of the print statement and e.printStackTrace().

File: src/main/java/com/google/sps/servlets/NewPostServlet.java [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+            e.printStackTrace(); 
+            throw new ServletException("Error: " + e.getMessage(), e);
+            
+        }
+        long repId = rep.getID();
+        //Insert comment and pull commentId
```
Comment (v1shanker) [inline src/main/java/com/google/sps/servlets/NewPostServlet.java:51]: these comments are extraneous

File: src/main/java/com/google/sps/servlets/RepAnswerServlet.java [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.ServletException;
+
+/*
```
Comment (v1shanker) [inline src/main/java/com/google/sps/servlets/RepAnswerServlet.java:21]: javadoc - /** */

File: src/main/java/com/google/sps/servlets/RepListServlet.java [hunk 26]
```diff
@@ -1,15 +1,83 @@
...
+            throw new ServletException("Error: " + e.getMessage(), e);
+        }
 
+        HttpEntity responseEntity = httpresponse.getEntity();
+        if(responseEntity != null) {
+            responseString = EntityUtils.toString(responseEntity);
+        }
+        else{
+            System.out.println("Response entity was null");
+            System.exit(0);
```
Comment (v1shanker) [inline src/main/java/com/google/sps/servlets/RepListServlet.java:76]: don't exit - throw an exception.
get rid of all print statements, etc.

File: src/main/java/com/google/sps/data/Comment.java (hunk 14 above)
Comment (jindaln) [reply src/main/java/com/google/sps/data/Comment.java:28]: doesn't check ID cause doesn't need to be the same entity in datastore - mainly for testing purposes

File: src/main/java/com/google/sps/data/Comment.java (hunk 15 above)
Comment (jindaln) [reply src/main/java/com/google/sps/data/Comment.java:30]: Swetha will take care of hashcode

File: .gitignore [hunk 27]
```diff
@@ -0,0 +1,4 @@
+.env
+/target
+*.class
+
```
Comment (v1shanker) [inline .gitignore:4]: don't need an extra blank line (not to be confused with trailing newline)

File: src/main/java/com/google/sps/data/Comment.java [hunk 28]
```diff
@@ -1,11 +1,48 @@
...
+            return false; 
+        } 
+        Comment that = (Comment) o;
+        return that.getComment().equals(this.comment)
+            && that.getDisplayName().equals(this.name);
+    } 
+
+    @Override
+    public int hashCode(){
+        //TODO
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:38]: this will have to return an int for it to compile...

you can return 42; for now. 

TODO's should be of the form 

TODO(username): explanation of the todo (such as implement hashCode().

File: src/main/java/com/google/sps/data/Constants.java [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+    public static final String POST_ENTITY_TYPE = "Post";
+    public static final String POST_QUESTION = "Question";
+    public static final String POST_ANSWER = "Answer";
+    public static final String POST_REPLIES = "Replies";
+
+    public static final String COMMENT_ENTITY_TYPE = "Comment";
+    public static final String COMMENT_NAME = "Nick Name";
+    public static final String COMMENT_MSG = "Message";
+
+    private static final Logger logger = LogManager.getLogger("Errors");
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Constants.java:18]: as mentioned in the other PR, please make the logger a private static variable in each class where the TAG is the class name.

File: src/main/java/com/google/sps/data/InsertAndUpdate.java [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key; 
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import java.util.ArrayList;
+import java.util.List;
+
+public class InsertAndUpdate{
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/InsertAndUpdate.java:17]: I had talked to Swetha about a better way to structure and name these classes (InsertAndUpdate, Parse, and Query). Is she going to address that in another PR. If so, please leave a TODO.

File: src/main/java/com/google/sps/data/Post.java (hunk 20 above)
Comment (v1shanker) [reply src/main/java/com/google/sps/data/Post.java:30]: please address this.

File: src/main/java/com/google/sps/data/Post.java [hunk 31]
```diff
@@ -1,15 +1,68 @@
...
+         
+        Post that = (Post) o;
+        return that.getQuestion().equals(this.question) && 
+               that.getAnswer().equals(this.answer) && 
+               that.getReplies().equals(this.replies); 
+    } 
+
+    @Override
+    public int hashCode(){
+        //TODO
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Post.java:51]: this will need to return an int to compile.

you can do a return 42; for now.

Also please leave a TODO in the format 

// TODO(username): explanation of what needs to be done.

File: src/main/java/com/google/sps/data/Post.java [hunk 32]
```diff
@@ -1,15 +1,68 @@
...
+    }
+
+    @Override 
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("Post: ").append(System.getProperty("line.separator")); 
+        sb.append("Question: ").append(System.getProperty("line.separator")); 
+        sb.append(question.toString()); 
+        sb.append("Answer: ").append(System.getProperty("line.separator")); 
+        sb = (answer == null) ? sb.append("") : sb.append(answer.toString());
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Post.java:61]: nit: !answer instead of answer == null

File: src/main/java/com/google/sps/data/QueryDatastore.java [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key; 
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/QueryDatastore.java:16]: javadoc?

File: src/main/java/com/google/sps/data/Parse.java [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key; 
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Parse.java:16]: javadoc?

File: src/main/java/com/google/sps/data/InsertAndUpdate.java (hunk 30 above)
Comment (v1shanker) [inline src/main/java/com/google/sps/data/InsertAndUpdate.java:17]: javadoc?

File: src/main/java/com/google/sps/servlets/NewPostServlet.java [hunk 35]
```diff
@@ -0,0 +1,63 @@
...
+and updates the list of posts associated with a representative when a user enters a 
+question on a representative's feed
+*/
+
+@WebServlet ("/new_post")
+public class NewPostServlet extends HttpServlet{    
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException{
+        String repName = request.getParameter("repName");
```
Comment (v1shanker) [inline src/main/java/com/google/sps/servlets/NewPostServlet.java:35]: use constants for these. Either private static variables in the class or in Constants.java

File: src/main/java/com/google/sps/servlets/NewPostServlet.java [hunk 36]
```diff
@@ -0,0 +1,63 @@
...
+        String name = request.getParameter("name");
+        String comment = request.getParameter("comment");
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        Entity repEntity = QueryDatastore.queryForRepresentative(repName); 
+        Representative rep = null;
```
Comment (v1shanker) [inline src/main/java/com/google/sps/servlets/NewPostServlet.java:41]: nit: don't need = null. its the default.

File: src/main/java/com/google/sps/servlets/NewPostServlet.java [hunk 37]
```diff
@@ -0,0 +1,63 @@
...
+
+        Entity repEntity = QueryDatastore.queryForRepresentative(repName); 
+        Representative rep = null; 
+        try {
+            rep = Parse.parseRepresentative(repEntity); 
+        } 
+        catch(EntityNotFoundException e) {
+            Constants.logger.error(e);
+            throw new ServletException("Error: " + e.getMessage(), e);
+
```
Comment (v1shanker) [inline src/main/java/com/google/sps/servlets/NewPostServlet.java:48]: extra line.

File: src/main/java/com/google/sps/servlets/RepAnswerServlet.java [hunk 38]
```diff
@@ -0,0 +1,47 @@
...
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.ServletException;
+
+/**
+The RepAnswerServlet class inserts a comment entity into datastore and updates
+the answer property of the post that the representative is responding to
+*/
+
+@WebServlet ("/rep_answer")
+public class RepAnswerServlet extends HttpServlet{
```
Comment (v1shanker) [inline src/main/java/com/google/sps/servlets/RepAnswerServlet.java:27]: comments from the new post servlet apply here.

File: src/main/java/com/google/sps/servlets/RepresentativeListServlet.java [hunk 39]
```diff
@@ -0,0 +1,79 @@
...
+
+/** 
+The RepListServlet class takes in an address from the user and call the Google Civic
+Info API to pull the list of representatives relevant for that zipcode and returns
+a json formatted objects which contains corresponding offices and officials
+*/
+
+@WebServlet ("/rep_list")
+public class RepresentativeListServlet extends HttpServlet{
+    Dotenv dotenv = Dotenv.load();
```
Comment (v1shanker) [inline src/main/java/com/google/sps/servlets/RepresentativeListServlet.java:33]: Same comments from PR#2 apply to this PR as well. Ideally you merge the other one first than revert the changes to this file.

File: src/main/webapp/script.js [hunk 40]
```diff
@@ -12,17 +12,64 @@
...
+}
 
-  // Add it to the page.
-  const greetingContainer = document.getElementById('greeting-container');
-  greetingContainer.innerText = greeting;
+//Makes call to repInDatastoreServlet to check if rep has made an account
+function checkIfRepInDatastore(repName){
+    fetch(`/rep_in_datastore?repName=${repName}`).then(response => {return Boolean.parseBoolean(response)});
 }
+
```
Comment (v1shanker) [inline src/main/webapp/script.js:75]: have a trailing newline at the end of the file, but not a blank line.

File: src/main/java/com/google/sps/data/Post.java (hunk 32 above)
Comment (jindaln) [reply src/main/java/com/google/sps/data/Post.java:61]: !answer won't work - maven says bad operand type Comment for unary operator

File: src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java [hunk 41]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * Converts entities from datastore into objects 
+ */  
+public class DatastoreEntityToObjectConverter {
+    /**
+     * Converts a representative entity into a representative object 
+     * @param entity of the representative 
+     * @return the representative object 
+     */  
+    protected static Representative convertRepresentative(Entity entity) throws EntityNotFoundException{
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java:27]: move the thows EntityNotFoundException to a new line so it doesn't wrap.

generally if a line is > 100 chars, break it into multiple.

File: src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java [hunk 42]
```diff
@@ -0,0 +1,100 @@
...
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Key; 
+import com.google.appengine.api.datastore.KeyFactory;
+
+/**
+ * Converts entities from datastore into objects 
+ */  
+public class DatastoreEntityToObjectConverter {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java:21]: can this class be final or do you mock it somewhere?

File: src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java [hunk 43]
```diff
@@ -0,0 +1,100 @@
...
+     * Converts a representative entity into a representative object 
+     * @param entity of the representative 
+     * @return the representative object 
+     */  
+    protected static Representative convertRepresentative(Entity entity) throws EntityNotFoundException{
+        String name = (String) entity.getProperty(Constants.REP_NAME);
+        String title = (String) entity.getProperty(Constants.REP_TITLE);
+        List<Post> posts = convertPostsFromRep(entity); 
+        long id = entity.getKey().getId();
+        System.out.println("Reached convertrep");
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java:32]: remove

File: src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java [hunk 44]
```diff
@@ -0,0 +1,100 @@
...
+        return new Representative(name, title, posts, id);   
+    }
+
+    private static List<Post> convertPostsFromRep(Entity repEntity) throws EntityNotFoundException{
+        List<Long> postIds = (ArrayList<Long>) repEntity.getProperty(Constants.REP_POSTS); 
+        System.out.println("THe postids: " + postIds);
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java:38]: remove

File: src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java [hunk 45]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        for (long postId : postIds) {
+            Key postEntityKey = KeyFactory.createKey(Constants.POST_ENTITY_TYPE, postId);
+            Entity postEntity = (Entity) datastore.get(postEntityKey); 
+            Post post = convertPost(postEntity); 
+            System.out.println("This is the post: " + post);
+            posts.add(post); 
+        }
+        System.out.println("These are the posts: " + posts);
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java:51]: remove

File: src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java [hunk 46]
```diff
@@ -0,0 +1,100 @@
...
+        return posts; 
+    }
+
+    protected static Post convertPost(Entity postEntity) throws EntityNotFoundException{
+        long questionId = (long) (postEntity.getProperty(Constants.POST_QUESTION));
+        Comment question = convertComment(questionId);
+        System.out.println("This is the question: " + question);
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreEntityToObjectConverter.java:58]: please remove all the print statements.

File: src/main/java/com/google/sps/data/DatastoreManager.java [hunk 47]
```diff
@@ -0,0 +1,192 @@
...
+        ds.put(postEntity); 
+        return postEntity.getKey().getId(); 
+    }
+
+    /**
+     * Updates a representative entity with a new post entity 
+     * @param repId ID of the representative entity 
+     * @param postId ID of the post entity 
+     */ 
+    public static void updateRepresentativePostList(long repId, long postId) throws EntityNotFoundException {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreManager.java:77]: break the line into two since it wraps. here and everywhere else.

File: src/main/java/com/google/sps/data/DatastoreManager.java [hunk 48]
```diff
@@ -0,0 +1,192 @@
...
+        datastore.put(postEntity);
+    }
+
+    /**
+     * Searches datastore for a particular representative and converts the entity 
+     * to a Representative instance 
+     * @param repName name of the representative to search datastore for 
+     * @return the representative object, or null if it was not found in datastore 
+     */ 
+    public static @Nullable Representative queryForRepresentativeObjectWithName(String repName) {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreManager.java:125]: you aren't consistent with the use of Nullable. While I prefer it, i'm ok if you get rid of it. just be consistent otherwise it isn't useful.

File: src/main/java/com/google/sps/data/DatastoreManager.java [hunk 49]
```diff
@@ -0,0 +1,192 @@
...
+            return null; 
+        }
+        else {
+            Representative rep = null; 
+            try {
+                rep = DatastoreEntityToObjectConverter.convertRepresentative(repEntity); 
+                return rep; 
+            } 
+            catch(EntityNotFoundException e) {
+                e.printStackTrace();
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreManager.java:146]: use the logger. not printStrackTrace.

you don't have access to system.out in prod generally.

File: src/main/java/com/google/sps/data/Post.java [hunk 50]
```diff
@@ -1,15 +1,68 @@
 package com.google.sps.data;
 
+import java.util.ArrayList;
 import java.util.List;
+import final com.google.sps.data.Comment;
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Post.java:5]: get rid of the final

File: src/main/java/com/google/sps/data/Representative.java [hunk 51]
```diff
@@ -1,16 +1,65 @@
...
+        return 42;
+        //TODO(swetha-gangu): implement hashCode()
+    }
+
+    @Override 
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("Representative: ").append(System.getProperty("line.separator")); 
+        sb.append("Representative name: ").append(name).append(System.getProperty("line.separator")); 
+        sb.append("Representative title: ").append(title).append(System.getProperty("line.separator"));
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Representative.java:59]: weird line break.
